mixer/pkg/runtime/dispatcher/util.go

```diff
@@ -15,54 +15,31 @@
 package dispatcher
 
 import (
-	"fmt"
-
 	"istio.io/istio/mixer/pkg/attribute"
 )
 
-// getIdentityAttributeValue from the attribute bag, based on the id attribute.
-func getIdentityAttributeValue(attrs attribute.Bag, idAttribute string) (string, error) {
-
-	v, ok := attrs.Get(idAttribute)
-	if !ok {
-		return "", fmt.Errorf("identity parameter not found: '%s'", idAttribute)
-	}
-
-	var destination string
-	if destination, ok = v.(string); !ok {
-		return "", fmt.Errorf("identity parameter is not a string: '%s'", idAttribute)
-	}
-
-	return destination, nil
-}
-
-// getNamespace returns the namespace portion of a given fully qualified destination name. If the destination
-// namespace cannot be deduced, then empty string is returned.
-func getNamespace(destination string) string {
-	l := len(destination)
-
-	idx1 := -1
-	for i := 0; i < l; i++ {
-		if destination[i] == '.' {
-			idx1 = i
-			break
+// getIdentityNamespace returns the namespace scope for the attribute bag
+func getIdentityNamespace(attrs attribute.Bag) (namespace string, err error) {
```

Comment:
This both hardwires the identity attribute, and further bakes in the namespace concept in the Mixer code. With MCP protocol based config model we have the opportunity to perform namespace based scoping/resolutions at a higher level and reduce the specialized logic in Mixer's machinery. 

Can we retain the original structure?

Reply:
I think the function may be badly named, but there is no longer identity attribute. We are assigning every request to a namespace, since namespace is the best thing we can do without service granularity (and destination service is deprecated, confusing, not well defined, or missing, as discussed many times over). Mixer needs to know the namespace, as it receives request from multiple namespaces, so not sure what you mean by "baking" namespace concept. We should be careful about isolating and segmenting policies.

Can you clarify your comment? I don't know MCP is.

Reply:
I don't understand why we don't need the identity attribute. IdentityAttribute is set up-top, when Mixer is started and doesn't get changed. The point of that attribute is to avoid hard-wiring, which this change is doing. It doesn't seem like it is a necessary change.

Mixer performs two namespace-related operations: coalesce and categorize rules based on the namespace they are declared in, and parse the identity attribute, assuming it is a FQDN, to detect the ruleset to apply. Otherwise, the rest of the code should be namespace agnostic.

With Galley, initially we will transfer the configuration resources as is, so Mixer still needs to perform ruleset coalescing, categorization based on namespaces (i.e. it needs to assume the resource name is FQN and parse it). However, Galley can do these namespace scoping operations in a general way (the pattern occurs in multiple places) and Mixer can stop doing this. That leaves only the rule-set selection mechanism to be namespace resolution dependent. Instead of further baking in the notion of a namespace in the rule, if we could go in the other direction (i.e. a simple equality match for example), then Mixer code will evolve towards just being a rule processing & transformation engine and get simplified.

Reply:
@ozevren I think some of that is valid, but the intent here is to generalize "identity attribute" to be an expression "if local use destination namespace, else source namespace". Unfortunately, the expression language is too limited for this, so I had to do it in code instead. There is nothing preventing that to move to an expression at a later time.

I don't mind namespace scoping being delegated to a lower level. The change is prior to actual rule selection, it's just extracting namespace in a different way.

We cannot be completely opaque about attributes and pass them blindly to galley. There is a major piece missing authenticating the source and authorizing access to policies based on source.  Namespace is easy to authenticate, not so much destination service.

Reply:
It boils down to the fact Mixer is used both for inbound and outbound traffic. For outbound traffic, we need to look at the config associated with the source workload, while for inbound traffic we need to look at the config associated with the destination workload.

We need an expedient way to address this need. As Galley develops, we can and will change this protocol.